File: uva_11572.py
def lis(s):
    if not len(s):
        return 0
    pos=[-1 for _ in range(1000001)]
    start, MAX=0, 0
    s.append(s[-1])
    for i in range(len(s)):
        if pos[s[i]]>=start:
            tmp=i-start
            MAX=MAX if MAX>tmp else tmp
            start=pos[s[i]]+1
            pos[s[i]]=i
        else:
            pos[s[i]]=i
    return MAX

case=int(input())
while case:
    case-=1
    n=int(input())
    data=[]
    for _ in range(n):
        data.append(int(input()))
    print(lis(data))

File: uva_11572.cpp
# include <iostream>
# include <vector>
using namespace std;
int lis(vector<int> s){
    if (!s.size())
        return 0;
    int pos[1000001];
    for (int i=0; i<1000001; ++i)
        pos[i]=-1;
    int start=0, MAX=0;
    s.push_back(s.back());
    for(int i=0; i<s.size(); ++i){
        if(pos[s[i]]>=start){
            int tmp=i-start;
            MAX = max(tmp, MAX);
            start = pos[s[i]]+1;
            pos[s[i]]=i;
        }else
            pos[s[i]]=i;
    }
    return MAX;
}
int main(void){
    int Case;
    cin >> Case;
    while(Case--){
        int n;
        cin >> n;
        vector<int> data;
        data.resize(n);
        for (int i=0; i<n; ++i)
            cin >> data[i];
        cout << lis(data) << endl;
    }
    return 0;
}